### PR TITLE
Rect2::has_point() breaks with negative size fixed

### LINKED ISSUE #37617: Rect2D .has_point() breaks with negative size values
**Godot version:**
3.2
**OS/device including version:**
MacOS Catalina 10.5.3

**Issue description:**
Rectangles .has_point() returns false when it should return true. In my case, I'm calculating a rectangle based on two points. This is done with subtraction, which results in the rectangle having negative size when one of the points is above or to the left of the other.

**Minimal reproduction project:**
```
var rect = Rect2D(Vector2(100, 100), Vector2(-100, -100));
if rect.has_point(Vector2(50, 50)):
    print("This should print");
else:
    print("But this prints instead");
```

### PR DESCRIPTION
**Pull request by ThakeeNathees**
Fix: #37617

*Bugsquad edit:* Mutually exclusive with at least a part of #37626

### CODE REVIEW

File: core/math/rect2.h [hunk 1]
```diff
@@ -152,14 +152,23 @@ struct Rect2 {
 		return new_rect;
 	};
 	inline bool has_point(const Point2 &p_point) const {
-		if (p_point.x < position.x)
+		Point2 _position;
+		Size2 _size;
+
+		_position.x = MIN(position.x, position.x + size.x);
```
Comment (groud) [inline core/math/rect2.h:158]: All this code could simply be replaced by a call to `Rect2.abs()` on a copy of the rect. And then the remaining code could be left intact if ran on the copy instead.

Comment (ThakeeNathees) [reply core/math/rect2.h:158]: @groud I've simplified the code with `abs()` and the not modifying the negative size ( unless there is no way to tell it the rect is flipped )

**Review by akien-mga — APPROVED** [review]

**Review by groud — COMMENTED** [review]
My only question left on this PR is likely the cost of creating a "duplicate" abs() rect every time. Maybe we should add a check on the size signs first to allow the creation of a new object only when required ?

Anyway, this kind of fix is, for me, preferred over #37626. The documentation might be updated though, to explain what a negative size means on such Rect2.

**Comment by akien-mga** [comment]
> My only question left on this PR is likely the cost of creating a "duplicate" abs() rect every time. Maybe we should add a check on the size signs first to allow the creation of a new object only when required ?

I guess we could cache the abs rect as an alternative representation of the negative sized rect, and use it in the methods that wouldn't work with negative size. But that depends on what the final consensus on #37626 would be indeed.

**Comment by aaronfranke** [comment]
I think it would be better to require the user to write `.abs().has_point()`. Doing this automatically can be bad for performance, and it does things behind the user's back (I know @reduz hates doing things behind the user's back).

Therefore, this should either be a debug warning (as I propose in #37626), or stated in the documentation.

**Comment by groud** [comment]
> Doing this automatically can be bad for performance, and it does things behind the user's back (I know @reduz hates doing things behind the user's back).

I disagree on that, it's not doing something in the user back. For me, a Rect2d with a negative size is a perfectly valid use case for which all API functions should work correctly. Whether or not we use abs() internally to implement the function or not is not important, as it does not change anything to the end user.

**Comment by reduz** [comment]
This is an interesting discussion. I am more in favor of thinking that if you use negative size values, has_point should always return false, so I think the current behavior is the expected one.

**Comment by mhilbrunner** [comment]
Closed, we discussed in a PR meeting and decided to go ahead with https://github.com/godotengine/godot/pull/37626 instead.

**Pull request closed without merging**